frontend/components/widgets/charts/sankey/component.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { FC } from 'react';
+import uniqBy from 'lodash/uniqBy';
+import { Tooltip, Sankey, ResponsiveContainer } from 'recharts';
+import SankeyLink from './sankey-link';
+import SankeyNode from './sankey-node';
+import { colors } from 'constants/charts';
+
+const COLORS = colors;
+
+export interface ConfigProps {
+  chartConfig: any;
+  tooltip: any;
+  sourceKey: string;
+  targetKey: string;
+  valueKey: string;
+  sourceColors?: string[];
+  targetColors?: string[];
+  colorLinksBy?: 'source' | 'target';
+}
+
+export interface ChartProps {
+  data: any[];
+  config: ConfigProps;
+}
+
+interface PrepareForSankeyArgs {
+  rawData: any[];
+  sourceKey: string;
+  targetKey: string;
+  valueKey: string;
+  sourceColors?: string[];
+  targetColors?: string[];
+  colorLinksBy?: 'source' | 'target';
+}
+
+interface PrepareForSankeyResult {
+  nodes: any[];
+  links: any[];
+}
+
+function prepareForSankey({
+  rawData,
+  sourceKey,
+  targetKey,
+  valueKey,
+  sourceColors = COLORS,
+  targetColors = COLORS,
+  colorLinksBy = 'source',
+}: PrepareForSankeyArgs): PrepareForSankeyResult {
+  const sources = uniqBy(
+    rawData.map((x, i) => ({ name: x[sourceKey], color: sourceColors[i % sourceColors.length] })),
+    'name',
+  );
+  const targets = uniqBy(
+    rawData.map((x, i) => ({ name: x[targetKey], color: targetColors[i % targetColors.length] })),
+    'name',
+  );
+
+  const nodes = sources.concat(targets);
+  const nodeNames = nodes.map((x: any) => x.name);
+
+  const links = rawData
+    .map((d: any) => ({
+      source: nodeNames.indexOf(d[sourceKey]),
+      target: nodeNames.indexOf(d[targetKey]),
+      value: d[valueKey],
+    }))
+    .map((d: any) => ({
+      ...d,
+      color: nodes[d[colorLinksBy]].color,
+    }));
+
+  return {
+    nodes,
+    links,
+  };
+}
+
+const Chart: FC<ChartProps> = ({ data, config }: ChartProps) => {
+  const { chartConfig, sourceKey, targetKey, valueKey, sourceColors, targetColors, colorLinksBy, tooltip } = config;
+  const chartData = prepareForSankey({
+    rawData: data,
+    sourceKey,
+    targetKey,
+    valueKey,
+    sourceColors,
+    targetColors,
+    colorLinksBy,
+  });
+
+  return (
+    <ResponsiveContainer width="100%" height={500}>
+      {/* @ts-expect-error: Disable type errors for SankeyLink and SankeyNode props missing */}
```

Comment:
Is this something we can fix? We control these components

Reply:
I don't know, I wouldn't bother now, this React way of `link={<SankeyLink />}` is a problem as Typescript cries that SankeyLink is not having any props. I silenced Typescript here, but I could do maybe some another function and explicitly pass props, but that won't be elegant. I prefer React way, and the way that is in documentation.